akka-http-core/src/main/scala/akka/http/scaladsl/model/ContentType.scala

```diff
@@ -31,74 +34,82 @@ object ContentTypeRange {
-          case MediaType.Encoding.Fixed(`cs`) ⇒ None
-          case x ⇒ throw new IllegalArgumentException(
-            s"MediaType $mediaType has a $x encoding and doesn't allow a custom `charset` $cs")
-        }
-      }
-    new ContentType(mediaType, definedCharset) {}
+  sealed trait NonBinary extends jm.ContentType.NonBinary with ContentType {
+    def binary = false
+    def charset: HttpCharset
+    def charsetOption = Some(charset)
   }
 
+  final case class WithFixedCharset(val mediaType: MediaType.WithFixedCharset)
+    extends jm.ContentType.WithFixedCharset with NonBinary {
+    def charset = mediaType.charset
+  }
+
+  final case class WithCharset(val mediaType: MediaType.WithOpenCharset, val charset: HttpCharset)
+    extends jm.ContentType.WithCharset with NonBinary {
+
+    private[http] override def render[R <: Rendering](r: R): r.type =
+      super.render(r) ~~ ContentType.`; charset=` ~~ charset
+  }
+
+  implicit def apply(mediaType: MediaType.Binary): Binary = Binary(mediaType)
+  implicit def apply(mediaType: MediaType.WithFixedCharset): WithFixedCharset = WithFixedCharset(mediaType)
+  def apply(mediaType: MediaType.WithOpenCharset, charset: HttpCharset): WithCharset = WithCharset(mediaType, charset)
+  def apply(mediaType: MediaType, charset: () ⇒ HttpCharset): ContentType =
+    mediaType match {
+      case x: MediaType.Binary           ⇒ ContentType(x)
+      case x: MediaType.WithFixedCharset ⇒ ContentType(x)
+      case x: MediaType.WithOpenCharset  ⇒ ContentType(x, charset())
+    }
+
+  def unapply(contentType: ContentType): Option[(MediaType, Option[HttpCharset])] =
+    Some(contentType.mediaType → contentType.charsetOption)
+
   /**
    * Tries to parse a ``ContentType`` value from the given String. Returns ``Right(contentType)`` if successful and
    * ``Left(errors)`` otherwise.
    */
   def parse(value: String): Either[List[ErrorInfo], ContentType] =
     headers.`Content-Type`.parseFromValueString(value).right.map(_.contentType)
+
+  private[http] case object `; charset=` extends SingletonValueRenderable
 }
 
 object ContentTypes {
   val `application/json` = ContentType(MediaTypes.`application/json`)
   val `application/octet-stream` = ContentType(MediaTypes.`application/octet-stream`)
-
-  val `text/plain` = ContentType(MediaTypes.`text/plain`)
-  val `text/plain(UTF-8)` = ContentType(MediaTypes.`text/plain`, HttpCharsets.`UTF-8`)
-  val `text/html` = ContentType(MediaTypes.`text/html`)
-  val `text/xml` = ContentType(MediaTypes.`text/xml`)
-
-  val `application/x-www-form-urlencoded` = ContentType(MediaTypes.`application/x-www-form-urlencoded`)
-  val `multipart/form-data` = ContentType(MediaTypes.`multipart/form-data`)
+  val `text/plain(UTF-8)` = MediaTypes.`text/plain` withCharset HttpCharsets.`UTF-8`
+  val `text/html(UTF-8)` = MediaTypes.`text/html` withCharset HttpCharsets.`UTF-8`
+  val `text/xml(UTF-8)` = MediaTypes.`text/xml` withCharset HttpCharsets.`UTF-8`
```

Comment:
Thought – so for making requests we certainly will want to specify UTF-8 anyway always, but when I'm using the low-level API, and pattern matching on a response... wouldn't there `text/xml` be useful, in case I really don't care what encoding it is in ("bah, ok windows-1250. but I just pipe it to a file anyway").?